fsw/cfe-core/src/es/cfe_es_apps.c

```diff
@@ -343,6 +343,236 @@ int32 CFE_ES_ParseFileEntry(const char **TokenList, uint32 NumTokens)
+**
+** Loads the module file via OSAL and stores all relevant info in the table entry as necessary.
+**
+**-------------------------------------------------------------------------------------
+*/
+int32 CFE_ES_LoadModule(const CFE_ES_ModuleLoadParams_t* LoadParams, CFE_ES_ModuleLoadStatus_t *LoadStatus)
+{
+    osal_id_t ModuleId;
+    cpuaddr StartAddr;
+    int32 ReturnCode;
+    int32 StatusCode;
+
+    StartAddr = 0;
+    ReturnCode = CFE_SUCCESS;
+
+    if (LoadParams->FileName[0] != 0)
+    {
+        /*
+         ** Load the module via OSAL.
+         */
+        StatusCode = OS_ModuleLoad ( &ModuleId,
+                LoadParams->Name,
+                LoadParams->FileName );
+
+        if (StatusCode != OS_SUCCESS)
+        {
+            CFE_ES_WriteToSysLog("ES Startup: Could not load file:%s. EC = 0x%08X\n",
+                    LoadParams->FileName, (unsigned int)StatusCode);
+            ModuleId = OS_OBJECT_ID_UNDEFINED;
+            ReturnCode = CFE_STATUS_EXTERNAL_RESOURCE_FAIL;
+        }
+    }
+    else
+    {
+        ModuleId = OS_OBJECT_ID_UNDEFINED;
+    }
+
+    /*
+     ** If the Load was OK, then lookup the address of the entry point
+     */
+    if (ReturnCode == CFE_SUCCESS && LoadParams->EntryPoint[0] != 0)
+    {
+        StatusCode = OS_SymbolLookup(&StartAddr, LoadParams->EntryPoint);
+        if (StatusCode != OS_SUCCESS)
+        {
+            CFE_ES_WriteToSysLog("ES Startup: Could not find symbol:%s. EC = 0x%08X\n",
+                    LoadParams->EntryPoint, (unsigned int)StatusCode);
+            ReturnCode = CFE_STATUS_EXTERNAL_RESOURCE_FAIL;
+        }
+    }
+
+    if ( ReturnCode == CFE_SUCCESS )
+    {
+        /* store the data in the app record after successful load+lookup */
+        LoadStatus->ModuleId = ModuleId;
+        LoadStatus->EntryAddress = StartAddr;
+    }
+    else if (OS_ObjectIdDefined(ModuleId))
+    {
+        /* If the module had been successfully loaded, then unload it,
+         * so that it does not consume resources */
+        StatusCode = OS_ModuleUnload(ModuleId);
+        if ( StatusCode != OS_SUCCESS ) /* There's not much we can do except notify */
+        {
+            CFE_ES_WriteToSysLog("ES Startup: Failed to unload: %s. EC = 0x%08X\n",
+                    LoadParams->Name, (unsigned int)StatusCode);
+        }
+    }
+
+    return ReturnCode;
+}
+
+/*
+**-------------------------------------------------------------------------------------
+** Name: CFE_ES_GetAppEntryPoint
+**
+** Helper function to act as the intermediate entry point of an app
+** This is to support starting apps before having a fully completed entry in the
+** global app table.  The app startup will delay until the app creation is completed
+** and verified, then the actual entry point will be determined.
+**
+**-------------------------------------------------------------------------------------
+*/
+int32 CFE_ES_GetAppEntryPoint(osal_task_entry *FuncPtr)
```

Comment:
Could this logic go in CFE_ES_RegisterApp to avoid having the entry point redirection?
What happens if the entry point cannot be found? Since this is running in the context of the new task, will it just end since it does not call the actual app entry point?
If so, should there be an else clause in CFE_ES_AppEntryPoint to clean up anything?

Reply:
Unfortunately that gives a chicken-egg situation.  `CFE_ES_RegisterApp()` is invoked by the app's init code, which is inside the entry point.  So this can't be used to find the entry point.

The reason for this is to impose a delay until the global table entries (app + task) are fully set up.  It just wants to see a coherent pair - a task entry which matches the running task context, which in turn refers to an app entry which refers back to the same task in its MainTaskID field.  Only once this positive-match criteria is all met will we "believe" the entry point field and call it.

Truth is - we should probably put the entry point on the task table, not the app table.  It would make this simpler because it only has to check coherency of the task table entry - although it couldn't be removed entirely, as the same fundamental issue would still exist. (i.e. can't set TaskID in global table until you know the ID, and you don't know the ID until the task is already running).

Reply:
And to answer your question - if the criteria is never met and an entry point is not known, the task will exit without starting anything.  If that ever happens though, it means there is a bug somewhere.